Porting ItemsRepeater from WinUI

grokys opened the pull request:
## What does the pull request do?

Ports the `ItemsRepeater` control and associated  infrastructure from the WinUI library: https://github.com/microsoft/microsoft-ui-xaml

The attached layout classes are added to `Avalonia.Layout`. The `ItemsRepeater` control and associated classes are in `Avalonia.Controls` (in a subdirectory).

I've not ported the `ItemsRepeater` tests because they're not super-easy to port and the range of functionality tested is rather basic.

Long-term it would be good to use `ItemsRepeater` as the presenter in `ListBox` etc, but this PR doesn't implement that.

## Not Ported in this PR

- `FlowLayout`
- `ScrollViewer`'s implementation of `IScrollAnchorProvider` is not implemented
- Animations (hopefully our existing animations system will suffice?)

## API Differences from UWP

- `ItemsRepeater.ItemsSource` is called `Items`
- `Layout` base class is called `AttachedLayout` because we already have `Layout` namespace

## Problems

Switching to horizontal stack layout in ControlCatalog causes a layout loop in the `ItemsRepeater` scrollbar, resulting in the scroll bar breaking. Need to investigate.

## Breaking changes

Moved `Avalonia.Controls.Orientation` enum to `Avalonia.Layout` namespace.

grokys committed:
Fix quick scrolling.

With a bit of a hack...

grokys committed:
Initial import of ItemsRepeater from the WinUI library.

https://github.com/microsoft/microsoft-ui-xaml

grokys committed:
Fix infinity checks.

Misread `std::isfinite` in C++ code as `std::infinite`, ha.

grokys committed:
Added UniformGridLayout.

ahopper commented:
Maximizing or resizing a window with a UniformGridLayout sometimes throws an IndexOutOfRange Exception. Apart from that it works very smoothly with a full screen of icons in my IconPacks app.

grokys committed:
Allow switching ItemsRepeater layout in ControlCatalog.

grokys committed:
Removed the ElementFactory args stuff.

We're just using `IDataTemplate` so it's not needed.

grokys committed:
Try up code.

Remove unused code and implement some focus stuff.

grokys committed:
Added headers.

grokys committed:
Added IScrollAnchorProvider.

Though nothing implements it yet. Uncomment/implement the parts of `ViewportManager` that required this interface.

grokys committed:
Scaffold IScrollAnchorProvider on ScrollViewer.

It doesn't do anything yet, but `ItemsPresenter` requires `ScrollViewer` to implement this interface.

grokys committed:
Added docs for ItemsRepeater-related classes.

grokys committed:
Moved attached layout code to Avalonia.Layout.

Also involved moving `Orientation` enum.

grokys committed:
Implemented NonVirtualizingLayout.

grokys committed:
More docs.

grokys committed:
Moved ItemsRepeater to Avalonia.Controls.

Keep it in a subdirectory though.

ahopper commented:
In https://github.com/ahopper/Avalonia.IconPacks/tree/xamlil
build 0.8.1-cibuild0003112-beta seems to run very slowly, as if it is not virtualizing.
This also repros the crash mentioned above if you scroll a little then maximise.

grokys committed:
Fix horizontal StackLayout.

grokys commented:
Thanks @ahopper, taking a look now.

grokys committed:
Merge branch 'master' into experimental/itemsrepeater

grokys committed:
Set default cache size to 0.

Scrolling on UWP happens asynchronous to the UI thread, so `ItemsRepeater` on that platform generate an additional buffer of prepared elements sufficient to fill an area larger than the viewport. We don't have async scroll rendering on Avalonia so make the default cache size 0.

grokys committed:
Avoid reset of containers when anchor element is replaced in ItemsRepeater.

* avoid reset of containers when anchor element is replaced

* fix element 0 ownership issue

Ported from https://github.com/microsoft/microsoft-ui-xaml/commit/e45e83d010502911aa1c7e14ea5bf681ceffb048

grokys committed:
Prevent IndexOutOfRange exception.

Prevent the exception described in https://github.com/AvaloniaUI/Avalonia/pull/2603#issuecomment-508011071. This makes our codebase diverge from WinUI's but Im unable to debug into the WinUI code consistently enough to work out why this problem doesn't show up there.

grokys commented:
@ahopper I've hopefully fixed those problems.

ahopper commented:
I can't make it crash anymore which is great, but it is still very slow, I need to check I've not done something silly since  my previous tests( when it was fast) but I I can't see anything on a quick look.

grokys committed:
Removed non-working code.

And replace it with a TODO.

grokys commented:
@ahopper ok yes I can reproduce the performance problem. Things are fast when the window is small, but if you maximize and then restore the window, the `ItemsRepeater` doesn't remove the extra items that were created when the window was maximized: they're left as children of the `ItemsRepeater` but off-screen and with size 0, waiting for recycling.

I think this falls into "things we need to optimize in Avalonia" rather than a problem with `ItemsRepeater`, so I think this PR is ready for proper review now.

kekekeks commented on src/Avalonia.Controls/IScrollAnchorProvider.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace Avalonia.Controls
+{
+    public interface IScrollAnchorProvider
```
This interface doesn't seem to be used anywhere, do we need it merged without implementation?

grokys replied:
I think we'll need to implement this soon, I left it in there but unimplemented so that I could at least port most of the code on the `ItemsRepeater` side that relates to scroll anchors.

kekekeks commented:
![изображение](https://user-images.githubusercontent.com/1067584/60668803-7c94ce80-9e75-11e9-96bb-9564f49de427.png)



-    switch to horizontal uniform grid
-    resize the window so it can accommodate two items
-    switch back to vertical stack
-    scroll to item 123
-    switch to horizontal uniform grid

grokys commented:
Yep, can reproduce if I add maximize the window before the last step.

grokys commented:
Actually all you need to do is to scroll down on the initial vertical and then switch to horizontal uniform grid.

This is caused by [this code](https://github.com/AvaloniaUI/Avalonia/blob/b26ec443749c8186c2d1b3aeed3f760280f56dd7/src/Avalonia.Layout/UniformGridLayoutState.cs#L68) creating `Item 1` to measure it. It then adds it to the flow algorithm, which doesn't seem to recycle it.

I suspect this is a bug in the WinUI `ItemsRepeater`, need to repro it there.

ahopper commented:
I agree the resize perf is just a nice to have.
MinItemWidth and Height do no trigger a redraw if changed by a binding 
```    <UniformGridLayout MinItemWidth="{Binding #iconsize.Value}" MinItemHeight="{Binding #iconsize.Value}"  />
      ```
I've added a slider to iconpacks with this a bound slider, the items only change size if you scroll or resize the view.

grokys commented:
@kekekeks ok, the problem seems to be caused by [this code](https://github.com/microsoft/microsoft-ui-xaml/pull/926/files#r300587386). I think I know how to fix it, but lets see if my understanding of the code is correct.

In the meantime I say we continue reviewing this PR and merge if everything else looks good, then I'll port the fix from WinUI in a later PR.

@ahopper I'll take a look at why changing those properties don't trigger a layout.

grokys committed:
Merge branch 'master' into experimental/itemsrepeater

grokys committed:
Realize a single element even if the repeater is outside the realized…

… range but still being measured.

Ported from https://github.com/microsoft/microsoft-ui-xaml/commit/9371dea6cc56857e277b189f4e093f68bc6e3b8e

grokys committed:
Undo unrelated change to StackPanel.

grokys committed:
Hack around problem with Item 0 being shown...

...when switching to a scrolled UniformGridLayout.

grokys commented:
@kekekeks I've hacked around that problem for the moment - I think this is ready for review/merge now.

grokys committed:
Merge branch 'master' into experimental/itemsrepeater

kekekeks commented:
Since it's a direct port from UWP, I don't think there is anything left for review. I think we need to remove the unused `IScrollAnchorProvider` and merge it as is.

grokys commented:
I think it would be best to leave `IScrollAnchorProvider` in and implement it on `ScrollViewer` in a separate PR.

kekekeks committed:
Merge branch 'master' into experimental/itemsrepeater

grokys commented on src/Avalonia.Layout/AttachedLayout.cs:
```diff
@@ -0,0 +1,106 @@
+    /// <summary>
+    /// Represents the base class for an object that sizes and arranges child elements for a host.
+    /// </summary>
+    public abstract class AttachedLayout : AvaloniaObject
```
Apologies, I missed this comment before. I wanted to make as few changes from the WinUI code as possible, and this is an abstract class in WinUI.
